### PR TITLE
[SYCLomatic] Add const qualifier for copy constructor of type passed througth kernel

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/lib/DPCT/Utility.cpp [hunk 1]
```diff
@@ -5047,6 +5048,22 @@ void checkTrivallyCopyable(QualType QT, clang::dpct::MigrationRule *Rule) {
         if (!C->isImplicit() && !C->isDeleted()) {
           if (C->isCopyConstructor()) {
             Messages.push_back("copy constructor");
+            // The 1st parameter of the copy constructor need "const" qualifier.
+            const ReferenceType *RT = dyn_cast<ReferenceType>(
+                C->getParamDecl(0)->getType().getTypePtr());
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: Need check if all copy ctor do not have const, then we can add const

File: clang/lib/DPCT/Utility.cpp [hunk 2]
```diff
@@ -19,6 +19,7 @@
 #include "clang/AST/ASTTypeTraits.h"
 #include "clang/AST/Expr.h"
 #include "clang/AST/ExprCXX.h"
+#include "clang/AST/Type.h"
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: remove

File: clang/lib/DPCT/Utility.cpp [hunk 3]
```diff
@@ -5047,6 +5048,22 @@ void checkTrivallyCopyable(QualType QT, clang::dpct::MigrationRule *Rule) {
         if (!C->isImplicit() && !C->isDeleted()) {
           if (C->isCopyConstructor()) {
             Messages.push_back("copy constructor");
+            // The 1st parameter of the copy constructor need "const" qualifier.
+            const ReferenceType *RT = dyn_cast<ReferenceType>(
+                C->getParamDecl(0)->getType().getTypePtr());
+            if (RT && !RT->getPointeeType().isConstQualified()) {
+              if (Rule) {
+                Rule->emplaceTransformation(new InsertText(
+                    C->getParamDecl(0)->getBeginLoc(), "const "));
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: auto Param0 = C->getParamDecl(0);

File: clang/lib/DPCT/Utility.cpp [hunk 4]
```diff
@@ -5047,6 +5048,22 @@ void checkTrivallyCopyable(QualType QT, clang::dpct::MigrationRule *Rule) {
...
+              } else {
+                auto FileNameAndOffset = DpctGlobalInfo::getLocInfo(
+                    C->getParamDecl(0)->getBeginLoc());
+                DpctGlobalInfo::getInstance().addReplacement(
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: InsertText.getreplacement();

File: clang/test/dpct/is_dev_copyable_spec.cu [hunk 5]
```diff
@@ -334,14 +334,14 @@ __global__ void k11(T) {}
 // CHECK-NEXT: */
 // CHECK-NEXT: struct B2 {
 // CHECK-NEXT:   B2() {}
-// CHECK-NEXT:   B2(B2& other) {}
```
Comment (zhiweij1) [inline clang/test/dpct/is_dev_copyable_spec.cu:337]: Add one more test like "B2(B2) {}"

File: clang/lib/DPCT/Utility.cpp (hunk 1 above)
Comment (zhiweij1) [reply clang/lib/DPCT/Utility.cpp]: ok

File: clang/lib/DPCT/Utility.cpp (hunk 3 above)
Comment (zhiweij1) [reply clang/lib/DPCT/Utility.cpp]: ok

File: clang/lib/DPCT/Utility.cpp (hunk 4 above)
Comment (zhiweij1) [reply clang/lib/DPCT/Utility.cpp]: ok

File: clang/test/dpct/is_dev_copyable_spec.cu (hunk 5 above)
Comment (zhiweij1) [reply clang/test/dpct/is_dev_copyable_spec.cu:337]: Not legal c++ syntax.

File: clang/lib/DPCT/Utility.cpp (hunk 2 above)
Comment (zhiweij1) [reply clang/lib/DPCT/Utility.cpp]: done

File: clang/lib/DPCT/Utility.cpp [hunk 6]
```diff
@@ -5043,15 +5043,52 @@ void checkTrivallyCopyable(QualType QT, clang::dpct::MigrationRule *Rule) {
...
       for (const auto &C : ClassDecl->ctors()) {
         if (!C->isImplicit() && !C->isDeleted()) {
           if (C->isCopyConstructor()) {
             Messages.push_back("copy constructor");
+            // The 1st parameter of the copy constructor need "const" qualifier.
+            const auto *FirstParam = C->getParamDecl(0);
+            const ReferenceType *RT =
+                dyn_cast<ReferenceType>(FirstParam->getType().getTypePtr());
+            if (RT) {
+              QualType PointeeTy = RT->getPointeeType();
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: use getQualifiers()

File: clang/lib/DPCT/Utility.cpp [hunk 7]
```diff
@@ -5043,15 +5043,45 @@ void checkTrivallyCopyable(QualType QT, clang::dpct::MigrationRule *Rule) {
...
             Messages.push_back("copy constructor");
+            // The 1st parameter of the copy constructor need "const" qualifier.
+            const auto *FirstParam = C->getParamDecl(0);
+            const ReferenceType *RT =
+                dyn_cast<ReferenceType>(FirstParam->getType().getTypePtr());
+            if (RT) {
+              Qualifiers Q = RT->getPointeeType().getQualifiers();
+              unsigned int CVQualifiers =
+                  Q.getCVRQualifiers() &
+                  (Qualifiers::Volatile | Qualifiers::Const);
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]:
```suggestion
;
```

File: clang/lib/DPCT/Utility.cpp [hunk 8]
```diff
@@ -5043,15 +5043,45 @@ void checkTrivallyCopyable(QualType QT, clang::dpct::MigrationRule *Rule) {
...
+              bool HasVolatile = CVQualifiers & Qualifiers::Volatile;
+              if (CVQualifiers & Qualifiers::Const)
+                CtorConstQualifierInsertLocations[HasVolatile].first = true;
+              else
+                CtorConstQualifierInsertLocations[HasVolatile].second =
+                    FirstParam->getBeginLoc();
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]:
```suggestion
              CtorConstQualifierInsertLocations[HasVolatile].first |= CVQualifiers & Qualifiers::Const;
              CtorConstQualifierInsertLocations[HasVolatile].second = FirstParam->getBeginLoc();
```

**Review by ziranzha — APPROVED** [review]

**Review by zhimingwang36 — APPROVED** [review]

**Pull request merged**